src/stream.rs

```diff
@@ -0,0 +1,301 @@
+                    return Async::NotReady
+                }
+                Ok(Async::Ready(item)) => match item {
+                    Some(Item::Data(body)) => {
+                        trace!("[{}] received data: {:?}", self.id, body);
+                        let body_len = body.bytes().len();
+                        self.buffer.extend(body.into_bytes());
+                        let remaining = self.recv_window.decrement(body_len);
+                        if remaining == 0 {
+                            trace!("[{}] received window exhausted", self.id);
+                            let item = Item::WindowUpdate(self.config.receive_window);
+                            if let Err(e) = self.send_item(item) {
+                                error!("[{}] failed to send window update: {}", self.id, e);
+                                self.state = State::Closed;
+                            }
+                            self.recv_window.set(self.config.receive_window as usize);
+                        }
+                        continue
+                    }
+                    Some(Item::WindowUpdate(n)) => {
+                        trace!("[{}] received window update: {}", self.id, n);
+                        self.send_window = self.send_window.checked_add(n).unwrap_or(u32::MAX);
+                        if let Some(task) = self.writer_task.take() {
+                            trace!("[{}] notifying writer task", self.id);
+                            task.notify()
+                        }
+                        continue
+                    }
+                    Some(Item::Finish) => {
+                        trace!("[{}] received finish", self.id);
+                        if self.state == State::SendClosed {
+                            self.state = State::Closed;
+                            return Async::Ready(())
+                        } else {
+                            self.state = State::RecvClosed
+                        }
+                        continue
+                    }
+                    Some(Item::Reset) => {
+                        trace!("[{}] received reset", self.id);
+                        self.state = State::Closed;
+                        return Async::Ready(())
+                    }
+                    None => {
+                        trace!("[{}] receiver returned None", self.id);
+                        self.state = State::Closed;
+                        return Async::Ready(())
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl io::Read for Stream {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if self.state == State::Closed || self.state == State::RecvClosed {
+            return Ok(0)
+        }
+        if self.poll_receiver().is_ready() {
+            return Ok(0)
+        }
+        if self.buffer.is_empty() {
+            return Err(io::Error::new(io::ErrorKind::WouldBlock, "not ready"))
```

Comment:
I'd prefer a `match` block, so that we clearly associate `Async::NotReady` to `WouldBlock`.
It is a logic error to return `WouldBlock` if a `NotReady` has not generated, so I think it's a good idea to clearly show the association in the source code.